File: src/Infrastructure/Proxirae.Infrastructure.ProcessCommunication/Pipes/Message/PipeMessage.cs
using System.Runtime.InteropServices;

namespace Proxirae.Infrastructure.ProcessCommunication.Pipes.Message
{
    public class PipeMessage
    {
        public PipeMessageHeader Header { get; set; }
        public byte[] Payload { get; set; } = [];

        public static PipeMessage Create(ushort type)
        {
            return new PipeMessage
            {
                Header = new PipeMessageHeader
                {
                    Type = type,
                }
            };
        }

        public static PipeMessage Create(ushort type, byte[] payload)
        {
            return new PipeMessage
            {
                Header = new PipeMessageHeader
                {
                    Type = type,
                },
                Payload = payload
            };
        }

        public static bool TryParse(ReadOnlySpan<byte> rawData, out PipeMessage? message)
        {
            message = null;
            int headerSize = Marshal.SizeOf<PipeMessageHeader>();

            if (rawData.Length < headerSize)
            {
                return false;
            }

            var header = MemoryMarshal.Read<PipeMessageHeader>(rawData[..headerSize]);

            if (header.Magic != 0x5850)
            {
                return false;
            }

            message = new PipeMessage { Header = header };

            if (header.PayloadSize > 0)
            {
                if (rawData.Length < headerSize + header.PayloadSize)
                {
                    return false;
                }

                message.Payload = rawData.Slice(headerSize, (int)header.PayloadSize).ToArray();
            }

            return true;
        }
    }
}

File: src/Network/Proxirae.Network/src/ipc/PipeMessage.cpp
#include "ipc/PipeMessage.h"

namespace Proxirae {
	PipeMessage PipeMessage::Create(PipeMessageType type)
	{
		return PipeMessage{
			.header = PipeMessageHeader{
				.type = type
			}
		};
	}

	PipeMessage PipeMessage::Create(PipeMessageType type, std::vector<std::byte> payload)
	{
		return PipeMessage{
			.header = PipeMessageHeader{
				.type = type
			},
			.payload = payload
		};
	}

	std::optional<PipeMessage> Proxirae::PipeMessage::TryParse(std::span<const std::byte> rawData)
	{
		if (rawData.size() < sizeof(PipeMessageHeader)) {
			return std::nullopt;
		}

		PipeMessage msg;
		std::memcpy(&msg.header, rawData.data(), sizeof(PipeMessageHeader));

		if (msg.header.magic != 0x5850) {
			return std::nullopt;
		}

		if (msg.header.payloadSize > 0) {
			if (rawData.size() < sizeof(PipeMessageHeader) + msg.header.payloadSize) {
				return std::nullopt;
			}

			auto payloadStart = rawData.begin() + sizeof(PipeMessageHeader);
			msg.payload.assign(payloadStart, payloadStart + msg.header.payloadSize);
		}

		return msg;
	}
}